### PR TITLE
[SPARK-8868] SqlSerializer2 can go into infinite loop when row consists only of NullType columns

### PR DESCRIPTION
**Pull request by yhuai**
https://issues.apache.org/jira/browse/SPARK-8868

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer2.scala [hunk 1]
```diff
@@ -179,10 +179,13 @@ private[sql] object SparkSqlSerializer2 {
 
   /**
    * Check if rows with the given schema can be serialized with ShuffleSerializer.
+   * Right now, we do not support schemata with complex types, UDTs, or all data types
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer2.scala]: Typo: schemata?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer2.scala [hunk 2]
```diff
@@ -179,10 +179,13 @@ private[sql] object SparkSqlSerializer2 {
...
+   * of fields are NullTypes.
    */
   def support(schema: Array[DataType]): Boolean = {
     if (schema == null) return true
 
+    var allNulTypes = true
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer2.scala]: Typo: Nul

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer2.scala [hunk 3]
```diff
@@ -190,12 +193,15 @@ private[sql] object SparkSqlSerializer2 {
         case array: ArrayType => return false
         case map: MapType => return false
         case struct: StructType => return false
-        case _ =>
+        case NullType => // Do nothing
+        case _ => if (allNulTypes) allNulTypes = false // This is not a NullType.
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer2.scala]: For completeness's sake, don't we also need to set this in the struct type etc cases above?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer2.scala [hunk 4]
```diff
@@ -179,23 +179,38 @@ private[sql] object SparkSqlSerializer2 {
...
+          if (allNullTypes) allNullTypes = false
+          return false
+        case map: MapType =>
+          if (allNullTypes) allNullTypes = false
+          return false
+        case struct: StructType =>
+          if (allNullTypes) allNullTypes = false
+          return false
         case _ =>
+          if (allNullTypes) allNullTypes = false
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkSqlSerializer2.scala]: pretty minor, but this is the same as just setting `allNullTypes = false`

File: sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlSerializer2Suite.scala [hunk 5]
```diff
@@ -57,6 +56,8 @@ class SparkSqlSerializer2DataTypeSuite extends SparkFunSuite {
   checkSupported(DecimalType(10, 5), isSupported = true)
   checkSupported(DecimalType.Unlimited, isSupported = true)
 
+  // If NullType is the only data type in the schema, we do not support it.
```
Comment (JoshRosen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlSerializer2Suite.scala]: For completeness, do we need a test to check that schemas that contain a NullType column and a column of some other supported type are supported?

Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/SparkSqlSerializer2Suite.scala]: oh, we have that test. The table `shuffle` used in `SparkSqlSerializer2Suite` contains all supported types and NullType.

**Comment by JoshRosen** [comment]
LGTM, so I'm going to merge this for master (1.5.0) and 1.4.2.  Thanks @yhuai!

**Pull request closed without merging**